module/state_synchronization/requester/execution_data_requester.go

```diff
@@ -0,0 +1,783 @@
+		}
+
+		err = e.processBlockSeals(block)
+		if err != nil {
+			// all errors indicate a problem fetching data from the db
+			ctx.Throw(fmt.Errorf("failed to process finalized block %s: %w", blockID, err))
+		}
+	}
+}
+
+// Will return an error if any of the seals processed are for a block that does not exist in the db
+func (e *executionDataRequesterImpl) processBlockSeals(block *flow.Block) error {
+	logger := e.log.With().Str("finalized_block_id", block.ID().String()).Logger()
+	logger.Debug().Msgf("checking %d seals in block", len(block.Payload.Seals))
+
+	if len(block.Payload.Seals) == 0 {
+		return nil
+	}
+
+	// First, find all seals in the block and sort them by height (ascending)
+
+	fetchRequests := []fetchRequest{}
+
+	for _, seal := range block.Payload.Seals {
+		sealedBlock, err := e.blocks.ByID(seal.BlockID)
+
+		if err != nil {
+			return fmt.Errorf("failed to lookup sealed block in protocol state db: %w", err)
+		}
+
+		logger.Debug().
+			Str("sealed_block_id", seal.BlockID.String()).
+			Uint64("sealed_block_height", sealedBlock.Header.Height).
+			Msg("checking sealed block")
+
+		fetchRequests = append(fetchRequests, fetchRequest{
+			blockID: seal.BlockID,
+			height:  sealedBlock.Header.Height,
+		})
+	}
+
+	logger.Debug().Msgf("found %d seals in block", len(fetchRequests))
+
+	sort.Slice(fetchRequests, func(i, j int) bool {
+		return fetchRequests[i].height < fetchRequests[j].height
+	})
+
+	// Next check if there is a gap between the lowest sealed block height and the last processed
+	// height, and record any that were missed. This allows us to catch up if we dropped some blocks
+	// because the the queue filled
+
+	last := e.notificationState.LastSealed()
+	if last < fetchRequests[0].height-1 {
+		missedHeights, err := e.missingBefore(last+1, fetchRequests[0].height)
+
+		if err != nil {
+			return fmt.Errorf("failed to lookup missing sealed block in protocol state db: %w", err)
+		}
+
+		fetchRequests = append(missedHeights, fetchRequests...)
+	}
+
+	// Finally, enqueue all blocks and notify workers
+
+	highestSealed := last
+
+	for _, request := range fetchRequests {
+		logger.Debug().Msgf("enqueueing fetch request for block %d", request.height)
+
+		accepted := e.fetchQueue.Push(request)
+		if !accepted {
+			// Silently drop requests if the queue is full. They will be backfilled when the next
+			// finalized block is processed. This takes advantage of the fact that seals are
+			// guranteed by hotstuff to be contiguous.
+			logger.Warn().
+				Str("sealed_block_id", request.blockID.String()).
+				Uint64("sealed_block_height", request.height).
+				Msg("fetch request queue is full")
+			e.notificationState.Sealed(highestSealed)
+			return nil
+		}
+
+		logger.Debug().Msgf("queue has %d requests", e.fetchQueue.Len())
+
+		highestSealed = request.height
+		e.fetchNotifier.Notify()
+	}
+
+	e.notificationState.Sealed(highestSealed)
```

Comment:
I think a better way to track all of this would be to just have the `finalizationProcessingLoop` keep track of the last block height it processed. Then, if it notices a gap, it can go get that block from the badger db.

In fact, I actually don't think we really need queues anywhere here. We can have a small buffered channel for the finalization queue to avoid blocking the finalization callback, and the finalization processing loop can just read from that channel.

Reply:
I've updated to backfill based on finalized blocks instead of seals